Allow config file to specify requires for filters

kbrock opened the pull request:
I like the rake task and will keep iterating on it.

I feel the filter in `config.yml` is more useful if a user can also say 'here is where you can find these filters'

This allows you to add a require block in the config to do that

(the rake task broke down when I wanted to serve up the assets. so I still need this, bundler, or `init.rb`

kbrock commented:
Yea, this is much better than the `init.rb` mod.

Anything special you want me to do before you want to merge?

jedcn commented on lib/reveal-ck/config.rb:
```diff
@@ -50,7 +50,8 @@ def filter_defaults
         'asset_root' => 'https://assets-cdn.github.com/images/icons/',
-        'base_url' => 'https://github.com'
+        'base_url' => 'https://github.com',
+        'requires' => []
```
I haven't played around with this yet, but I noticed this entry is `requires` (plural) but up on line 39 of create_slides_html.rb in this PR it's `config.require` (singular). Is that on purpose?

kbrock replied:
thanks. :octocat: FTW

Huh, wonder why this is working.

Sure enough, when I change to master, this stops working. :confused:

jedcn commented:
For posterity, and so that everything is in one place, can you add a `config.yml` to this PR with the entries for `requires` that you're successfully using?

kbrock commented:
A. it looks like my config file had `require:` in it, so that is why it was working. Great catch.
B. `README.md` is getting a bit long. You want in there right below the `config.yml` section?`
C. Testing this is a bit tricky. Would prefer to not say "make sure `require` is called", but the alternative is not so nice.

jedcn commented:
When it comes to documentation, I'm open to suggestions. I don't actively promote reveal-ck, so the fact that you discovered it and have used it means you've got more insight into the helpfulness / problems with the `README.md` and documentation than I do :smile:.

1. I think the `README.md` is long, and should be shortened (rather than lengthened).
2. I think most people don't get to tweaking the filters or making custom filters.

Putting those two ideas together, this could be a good entry in the [wiki][reveal-ck-wiki]-- possibly as part of a [FAQ][reveal-ck-wiki-faq] answer for "How do I use a custom filter?" or a new wiki entry dedicated to that topic. Let me know what you think?

I wonder if I can write a `*.feature` for this. I'll give it some thought. If there was a `.feature` entry, that would generate a page up on https://relishapp.com/jedcn/reveal-ck, and that page could be linked to from the faq/wiki.

[reveal-ck-wiki]: https://github.com/jedcn/reveal-ck/wiki
[reveal-ck-wiki-faq]: https://github.com/jedcn/reveal-ck/wiki/FAQ

---

When it comes to testing, I'd do something like this: https://github.com/jedcn/reveal-ck/commit/72dfdb58ee0712a8759fcc866f10800532c272d2, but I think that's *exactly* what you said you didn't want to do? Does the test in that commit do the trick? Is it violating some rule, or somehow offensive?

---

While I was in the code just now, I was thinking that perhaps the notion of "requires" could be more general than just for filters. To be explicit-- the entry in the `config.yml` isn't `requires_for_filters` instead it is just `requires`.

If it were the case, the same code you've written that loops through things to require [might end up over here in slides_builder.rb][slides-builder].

[slides-builder]: https://github.com/jedcn/reveal-ck/blob/master/lib/reveal-ck/builders/slides_builder.rb#L32

jedcn commented:
This is what I came up with re: `config.filters` and `config.requires`: https://github.com/jedcn/reveal-ck/commit/91ad679ebdd01a7829ecc7cd35174303b5040084

As an aside, I like the `*.feature" entries because they give me perspective into how things can be used (from the outside in), and they remind me about how the project is actually being used.

kbrock commented:
the `*.feature` is great. (code and docs portion)
I agree requires is more general. But if you are adding code, you'd probably have the original requires in the actual ruby.

You want me to move the requires code?

jedcn commented:
No need to move the code. I can't think of any situation where you'd need to require something earlier on. If I do, I'll move it myself.

Looking at my schedule today, I'll plan to merge and release up to rubygems tonight. I'd hope to get the rake task in, too (https://github.com/jedcn/reveal-ck/pull/51). That's pretty cool.

kbrock commented:
Thanks for features merged them

kbrock commented:
This is the only thing I need for now. Thanks for all the help.

jedcn commented:
v3.2.0 should include this change.

It has been published [here](https://rubygems.org/gems/reveal-ck/versions/3.2.0).

And this specific feature is documented [here](https://relishapp.com/jedcn/reveal-ck/docs/the-capabilities-of-config-yml#generating-slides-with-custom-filters-and-requires).

And the CHANGELOG is [here](https://github.com/jedcn/reveal-ck/blob/ceca5f3d347ea717f277a05985bd5b29b3331ec9/CHANGELOG.md).

Please let me know if you see any problems.

I'll review and merge the other PRs shortly, and we can fix any problems then.

Hope this helps.

Thanks again!

kbrock commented:
@jedcn I just added a filter (implemented a `#include` filter) and added the entry to require and filters, and it works like a charm.

Thanks so much for merging these features that you will probably not use

jedcn commented:
Hah! Good to know. It's true.. I haven't yet used this feature myself.

Glad it's working for you, though!
